server/handler.go

```diff
@@ -115,26 +152,40 @@ func (h *Handler) ComQuery(
 
 			r = nil
 			proccesedAtLeastOneBatch = true
-
 			continue
 		}
 
-		row, err := rows.Next()
-		if err != nil {
+		select {
+		case err := <-errchan:
 			if err == io.EOF {
-				break
+				break rowLoop
 			}
 
 			return err
-		}
+		case row := <-rowchan:
+			outputRow, err := rowToSQL(schema, row)
+			if err != nil {
+				close(quit)
+				break rowLoop
+			}
 
-		outputRow, err := rowToSQL(schema, row)
-		if err != nil {
-			return err
+			r.Rows = append(r.Rows, outputRow)
+			r.RowsAffected++
+		case <-timer.C:
+			if h.readTimeout != 0 && waitTime >= h.readTimeout {
+				// Return so Vitess can call the CloseConnection callback
+				err = RowTimeout.New()
+				close(quit)
+				break rowLoop
+			}
 		}
+		timer.Reset(waitTime)
+	}
+
+	timer.Stop()
```

Comment:
there are returns previous to this line. So, maybe it should be `defer timer.Stop()` just after its creation

Reply:
Done.